org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompletionEngineProviderDiscovery.java

```diff
@@ -0,0 +1,97 @@
+package org.eclipse.jdt.internal.core;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExtension;
+import org.eclipse.core.runtime.IExtensionPoint;
+import org.eclipse.core.runtime.ILog;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.internal.codeassist.CompletionEngine;
+import org.eclipse.jdt.internal.codeassist.ICompletionEngineProvider;
+
+class CompletionEngineProviderDiscovery {
+	private static final String SELECTED_SYSPROP = "ICompletionEngineProvider"; //$NON-NLS-1$
+	private static final String CODE_ASSIST_PROVIDER_EXTPOINT_ID = "completionEngineProvider" ; //$NON-NLS-1$
+	private static boolean ERROR_LOGGED = false;
+
+	private static String lastId;
+	private static IConfigurationElement lastExtension;
+
+	private static final Map<String, ICompletionEngineProvider> ENGINE_PROVIDER_CACHE = new HashMap<>();
+
+	public static ICompletionEngineProvider getInstance() {
+		String id = System.getProperty(SELECTED_SYSPROP);
+		IConfigurationElement configElement = getConfigurationElement(id);
+		lastId = id;
+		lastExtension = configElement;
+		if (configElement != null) {
+			try {
+				if (ENGINE_PROVIDER_CACHE.get(id) != null) {
+					return ENGINE_PROVIDER_CACHE.get(id);
+				}
+				Object executableExtension = configElement.createExecutableExtension("class"); //$NON-NLS-1$
+				if (executableExtension instanceof ICompletionEngineProvider icep) {
+					ENGINE_PROVIDER_CACHE.put(id, icep);
+					return icep;
+				}
+			} catch (CoreException e) {
+				if (!setErrorLogged()) {
+					ILog.get().error("Could not instantiate ICompletionEngineProvider: '" + id + "' with class: " + configElement.getAttribute("class"), e); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+				}
+			}
+		}
+		return CompletionEngine::new;
+	}
+
+	/**
+	 *
+	 * @param id
+	 * @return The extension element with the given id or <code>null</null> if not found
+	 */
+	private static IConfigurationElement getConfigurationElement(String id) {
+		if (id == null || id.isBlank()) {
+			return null;
+		}
+		if (lastExtension != null && Objects.equals(id, lastId)) {
+			return lastExtension;
+		}
+		IExtensionPoint extension = Platform.getExtensionRegistry().getExtensionPoint(JavaCore.PLUGIN_ID, CODE_ASSIST_PROVIDER_EXTPOINT_ID);
+		if (extension != null) {
+			IExtension[] extensions = extension.getExtensions();
+			for (IExtension ext : extensions) {
+				IConfigurationElement[] configElements = ext.getConfigurationElements();
+				for (final IConfigurationElement configElement : configElements) {
+					String elementId = configElement.getAttribute("id"); //$NON-NLS-1$
+					if (id.equals(elementId) && "resolver".equals(configElement.getName())) { //$NON-NLS-1$
+						return configElement;
+					}
+				}
+			}
+		}
```

Comment:
Just wanted to confirm. The extension point framework handles correctly starting the bundle that contains the contributed extension right ? That's why you didn't just directly do `Class.forName(...).getMethod(...).invoke(...)` on the system property right ? I'm asking because otherwise it seems simpler to bypass all this.

Reply:
Honestly this class started out as a copy of [CompilationUnitResolverDiscovery](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/ca77e4c56980ac52d225595273979dfd1316cfd4/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/CompilationUnitResolverDiscovery.java#L23). I don't fully understand it, but it was the reference I was working off of.

Reply:
Ok fair, if it's based off of that, and that was merged, then let's stick to that approach since obviously there hasn't been any opposition to it.